components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java

```diff
@@ -709,26 +687,104 @@ private boolean isDomainNameEmbeddedInAttributeValue(String filterOperation, Str
-            String domainName) throws CharonException {
+    private String[] filterUsers(Node node, int offset, int limit, String sortBy, String sortOrder, String domainName)
+            throws CharonException {
 
+        // Filter users when the domain is specified in the request.
         if (StringUtils.isNotEmpty(domainName)) {
-            return getFilteredUsernames(node, offset, limit, sortBy, sortOrder, domainName);
+            return getFilteredUsernames(createConditionForSingleAttributeFilter(domainName, node), offset, limit,
+                    sortBy, sortOrder, domainName);
         } else {
+            // Filter users when the domain is not set in the request. Then filter through multiple domains.
             String[] userStoreDomainNames = getDomainNames();
             ArrayList<String> filteredUsernames = new ArrayList<>();
             for (String userStoreDomainName : userStoreDomainNames) {
-                filteredUsernames.addAll(Arrays
-                        .asList(getFilteredUsernames(node, offset, limit, sortBy, sortOrder, userStoreDomainName)));
+
+                // Create filter condition for each domain.
+                Condition condition = createConditionForSingleAttributeFilter(domainName, node);
+
+                // Filter users for given condition and domain.
+                String[] userNames = getFilteredUsernames(condition, offset, limit, sortBy, sortOrder,
+                        userStoreDomainName);
+                if (userNames == null) {
+                    userNames = new String[0];
+                }
+
+                // Calculating new offset and limit parameters.
+                int numberOfFilteredUsers = userNames.length;
+                if (numberOfFilteredUsers <= 0 && offset > 1) {
+                    offset = calculateOffset(condition, offset, sortBy, sortOrder, domainName);
+                } else {
+                    // Returned user names size > 0 implies there are users in that domain which is larger than the offset.
+                    offset = 1;
+                    limit = calculateLimit(limit, numberOfFilteredUsers);
+                }
+                filteredUsernames.addAll(Arrays.asList(userNames));
+
+                // If the limit is changed then filtering needs to be stopped.
+                if (limit == 0)
+                    break;
             }
             return filteredUsernames.toArray(new String[0]);
         }
     }
 
+    /**
+     * Method to update the count(limit) when iterating a filter across all domains.
+     *
+     * @param limit                 Counting value (limit)
+     * @param numberOfFilteredUsers Amount of users filtered in the search
+     * @return Calculated new limit (count)
+     */
+    private int calculateLimit(int limit, int numberOfFilteredUsers) {
+        if (log.isDebugEnabled()) {
+            log.debug("Updating the original count to be used for the next domain search. ");
```

Comment:
shall we log the some context info here?

Reply:
Fixed:f999948935ff2081b9362c4cbe0d6e8f5ee238db